### ISSUE TITLE
priorities confuse me

### ISSUE THREAD

**Issue by TanninOne** [loot/loot#866]
I'm afraid this is probably going to be a rtfm question but I still can't reconcile the observed behaviour with the documentation.

So I test with a plugin that should have practically no relations with other plugins and give it a global priority of 10 (via user metadata). Since most other plugins have no explicit priority set I expect that plugin to go very late in the sort order.
In loot 0.10.3 this was the case, in loot 0.12.1 the plugin stays where it was.
If I set the global priority to 60 however it now gets loaded last, so it seems more like the fallback priority is ~50, not 0 because as far as I understand it there shouldn't be a difference between 10 vs 60 if everything else defaults to 0?

**Comment by WrinklyNinja** [comment]
Priorities confuse everyone, they're a bad solution to an ill-defined problem.

That said, this sounds like it's probably due to priority inheritance, maybe something else is causing the other plugins to get ~50? If you turn on debug logging (in the LOOT application) / call SetLoggingCallback() to handle trace messages (if using the API), LOOT will log detail about priority inheritance, which may shed some light on what it's doing.

**Comment by TanninOne** [comment]
Ooooh, ok, I think I'm starting to understand. In the debug log I get this:

> [12:14:56.325050] [trace]: Doing DFS for Eli_Breezehome.esp which has local priority 0 and global priority 60
> [12:14:56.325050] [trace]: Doing DFS for dummy.esl which has local priority 0 and global priority 50
> [12:14:56.326050] [trace]: Overriding global priority for SkyHUD.esp from 0 to 50

So loot sees the esl file as a light master and sets the priorities of all regular plugins to be at least the prio of the esl so they get loaded first?

So if there is an explicit global priority among masters then the global priority of regular esps is `min(esp_global_prio || 0, max(master_global_prios)))`?

Are there other rules that override the priorities of plugins?

Also, why is the esl a light master, it doesn't have the flag set in it's bitmask?

**Comment by Freso** [comment]
> Also, why is the esl a light master, it doesn't have the flag set in it's bitmask?

That sounds like it should be filed as a separate issue. :)

**Comment by WrinklyNinja** [comment]
Plugins inherit the highest priority of any plugins they would otherwise load after anyway. While it's not obvious, it prevents not obvious and more clearly wrong behaviour (I added it in to fix a few bug reports).

As for `.esl` plugins being light masters even without the flag set, AFAIK this is the correct behaviour: last I checked the game will "upgrade" plugin flags at runtime to match their file extension, so all .esms are master files, and all .esls are light masters.
